### DISCUSSION TITLE
Any example using glacier?

### DISCUSSION THREAD

**Discussion by bbigras** [rustic-rs/rustic#692] [Q&A] [answered]
Do I just need to use a glacier s3 bucket for cold and a regular s3 bucket for hot?

 There's more than 1 Glacier StorageClass, and it seems a normal s3 bucket can transition its file to glacier https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/s3_bucket#transition

**Comment by aawsome** [comment, marked as answer]
rustic doesn't support single repositories/buckets which have objects that can be in different "states" (something like this is not contained in the restic REST protocol; rclone therefore isn't able to provide this kind of information).

So, to use S3 glacier, you should use 2 buckets: A hot one which is always accessible and a cold one where all object should be transitioned/transferred directly to Glacier when they are stored. The "hot" objects are then in fact stored in both buckets, but this shouldn't be much overhead as it is only metadata which usually is less than 1% of the repo size. (The side effect is that the cold repo contains a full restic-compatible repository; if you warm up it completely, you can use it a standard repository within rustic and even restic).

TL;DR: I think you have to do the following:
1. To store data (e.g. `init` or `backup` command):
- Create a regular bucket and use it as `repo-hot`
- Create a bucket for `cold` and configure it to store data directly in Glacier (don't know if you can configure this in `rclone` or in AWS)

2. To retrieve data (e.g. `restore`):
- configure `warm-up-command` such that objects from the cold buckets are restored, see e.g https://docs.aws.amazon.com/AmazonS3/latest/userguide/restoring-objects.html. That is, the config file should look have a line like `warm-up-command  = 'aws s3api restore-object --bucket BUCKET --key path/data/%id --restore-request '{"Days":25,"GlacierJobParameters":{"Tier":"Standard"}}'  '`  (may need some adaption)
- configure a suitable `warm-up-wait` to ensure that the data is warmed up and can be accessed after that
- Note: All commands which need to read cold data now will warm this data up and wait for the given time before processing the data

3. removing cold data (`prune`)
This actually should work out of the box, but note that by default `prune` doesn't repack any cold pack file; instead it keeps them until the last blob within is no longer used.
- to do more repacking, have a look at the `repack-cacheable` option which does allow to repack cold pack files.
- Use the `keep-pack` option if you have some minimum holding duration, i.e. if you would get charged if you remove objects too early. This option ensures that only pack files older than the given time will be removed.

Once you get a working configuration, can please share it so other users can also use it?

**Comment by mprasil** [comment]
I'm linking it here because this came up when I initially investigated how to use Rustic with Glacier.

I have written two articles on the topic - [economic side of hosting rustic repo in AWS Glacier](https://kmh.prasil.info/posts/rustic-cold-storage-glacier-economics/) and the [technical side of migrating to Glacier, restoring and such](https://kmh.prasil.info/posts/rustic-cold-storage-glacier-migration-configuration/). It's really quite complex topic from both points of view and there are some sharp edges especially around restoring the backups.

 Hopefully it helps someone.

**Reply by aawsome** [reply]
@mprasil Great docu! And especially great to hear rustic works with AWS Glacier for you!

Some comments:
- I hope your given credentials are artificial or no longer valid :wink: 
- Note there is `options` for common options (hot and cold) and `options-hot`/`options-cold` which apply only to hot/cold parts (and also overwrite `options` if same option is given) . Using this you could simplify your config a bit.
- About pruning: Note that pruning is perfectly supported for hot/cold repositories. The default is to not repack data pack files, for obvious reasons. But if you decide to do so nevertheless, they will be warmed-up and then repacked. Also no longer needed packs are simply removed.
- About object lock, you might want to have a look at https://github.com/rustic-rs/rustic/discussions/1050. Note that this is all WIP (but what you described should already perfectly work - this is just to add more options). 
- We are thinking about using profiles as targets for `copy`. This would mean you would have to define a profile of your hot/cold repository (as you anyway did) and specify the profile name in the `[copy]` section. What do you think about this idea?

**Reply by mprasil** [reply]
Thank you for feedback!

>I hope your given credentials are artificial or no longer valid 😉

I'm pretty sure someone tried those already. 😄 I'm also pretty sure they are not even in valid format as I didn't make much effort generating realistic credentials. But thanks for that warning just in case. 

>Note there is options for common options (hot and cold)

I'll have to try that. The documentation was sparse and at some stage it didn't work as I expected so I opted for more explicit config. Does `rustic show-config` show _effective_ configuration to experiment a bit with the configuration inheritance?

> About pruning: Note that pruning is perfectly supported for hot/cold repositories. The default is to not repack data pack files, for obvious reasons. But if you decide to do so nevertheless, they will be warmed-up and then repacked. Also no longer needed packs are simply removed.

Yeah I found that much. I just haven't personally seen repositories where that would make financial sense for cold storage. In my experience it barely made any sense on regular storage. (Perhaps it's just nature of my backups)

> About object lock

I've noticed that discussion, it's definitely very interesting development. In homelab/small office setups with mildly critical data, versioning might be about as good with less of a chance of shooting yourself in the foot. The locking is double edged sword especially if you're approaching it from a standpoint where you think $11 saved per month is worthwhile thing. Hope that makes sense.

> We are thinking about using profiles as targets for copy. 

I think that would be neat in terms of documentation as you could say that copy target is another profile and you don't have to document another config structure. In practical terms I still think it wouldn't save much work as the copy is once-off thing (at least in my case) and having to juggle multiple S3 providers it still makes more sense preparing separate "migration" configs with credentials baked in that I wouldn't want to keep after I'm done. But from documentation standpoint it's for sure great idea to not have duplicate structures under targets.

One thing that I think might be worthwhile to add is for the `warm-up-command` to have some path alternative to the `%id` placeholder. As it is now I had to write the wrapper script just to provide object key as `--key` parameter. This is not much work in fairness, but it does require that whoever wrote the script understands restic repository structure. And in case you add more object _types_ in the future that might need warming up, (or you move the files around) you might have to introduce breaking change to this script. Whereas if there was for example `%path` placeholder with relative path of the object any changes in the repository structure wouldn't be breaking change for the `warm-up-command`. (and one wouldn't need to write the wrapper in the first place)

Having some extra `aws-style-warmup` where rustic itself would issue the restore-object calls would be best. The current approach is quite slow once you get to hundreds of thousands objects to restore as you need to run command for each. Something like:

```toml
[repository.options-cold]
aws-style-warmup = true
aws-style-warmup-days = 5
aws-style-warmup-tier = "Bulk"
```

Alternatively if we're to rely on external command to not introduce specific provider code into rustic. (or for less popular providers) There could be something like `warm-up-pipe = "some_command"` where the specified command is expected to read list of objects on the `stdin` and warm them up. This way the command can optimize on how it would do it effectively without spamming processes.

**Reply by aawsome** [reply]
> Does rustic show-config show effective configuration to experiment a bit with the configuration inheritance?

Yes, exactly. This is what this command is for. The output format is a bit verbose, though, but this was the easiest to implement

> One thing that I think might be worthwhile to add is for the warm-up-command to have some path alternative to the %id placeholder.

Thanks for the suggestion. I think we will add this with support for a `locking` command. For warming-up, the id is sufficient as we are anyway only talking about pack files. But yes, having a `%path` would be quite handy anyway.

About a more convenient warm-up: It might be useful to try other tools than the AWS CLI (which is written in python and has a very slow start-up). @kapitainsky mentioned the minio `mc` command, maybe it is also able to change storage classes?

A even better alternative would be better support within rustic, right. From an architectural sight, warm-up is just a method in the backend trait. So , this could be done in `rustic_backend` or directly in opendal. I've opened a opendal issue: https://github.com/apache/opendal/issues/4308.
We would however need to change the backend trait to support exact wait-time feedback from the storage.

**Reply by mprasil** [reply]
I don't think you can use `mc` for this sadly. Minio does not have that kind of functionality, so the cli tool doesn't have it either. (I'd love if someone proved me wrong here, but AFAIK that is the case)

Another interesting option (although a bit manual) is to use [S3 Batch Operation job](https://docs.aws.amazon.com/AmazonS3/latest/userguide/batch-ops-create-job.html). One would just configure `warm-up-command` to do something like:

```sh
echo "${BUCKET_NAME},${OBJECT_PATH}" >> restore.csv
```

You can then take that CSV and submit a batch operation. (not sure if this can be done via AWS SDK or if it's console-only thing) This won't save you the per-request cost, because AWS will simply do the requests on your behalf, but they are able to do many requests really fast and they will also generate a report at the end, so you could kick off rustic restore operation once all of the objects in the report are restored.

**Reply by kapitainsky** [reply]
Firstly - great blog posts about Glacier and rustic. Good read and solid info.  Things like these slowly fill in the gap in documented solutions and "how to-s" .

> Another interesting option (although a bit manual) is to use [S3 Batch Operation job](https://docs.aws.amazon.com/AmazonS3/latest/userguide/batch-ops-create-job.html). One would just configure `warm-up-command` to do something like:

`aws` cli is slow - this is a fact. I was actually looking at batch ops myself - but it was more from theoretical point of view - so only exist as my note to investigate it further when needed. With my data it is not a big deal if I have to wait 1h instead of 1 min to request objects restoration. But I can see that for people trying to restore millions of objects it can be only solution really. Maybe good subject for the next blog series post?:)

PS. Your other blog posts are also cool and worth reading. In my RSS feed already:)

**Reply by mprasil** [reply]
Thanks for feedback! Yeah I agree on the batch job. The waiting times with deep archive are in hours anyways, so the speed of restore requests hardly makes a difference. Which is why I also didn't look much into it.

With restore requests in thousands I'd expect reliability being another sharp edge. The restore script might need to be able to do some retries for the occasional failed request. At that stage more dedicated tool (like the Batch operation) might be better just from that standpoint.

**Reply by aawsome** [reply]
@mprasil Can we link your stories [here](https://rustic.cli.rs/docs/stories.html)?

**Reply by mprasil** [reply]
Yes, feel free to link it.

**Reply by aawsome** [reply]
Thanks, is now linked. 
And, BTW, the first article of rustic (that I am aware of) :tada:

**Reply by mprasil** [reply]
I think I need to post couple updates, but before I do I figured I'll check it with you @aawsome.
* The copied repo has similar size, but about 10x _fewer_ pack files. (so going from ~300k -> ~30k) Is that down to bigger pack sizes? I have copied over all snapshots and the data seems to be there. (the overall repo size is similar)
  * This seems to save on upfront cost quite a bit. So nice side effect, just making sure it's expected.
* All of the snapshots that were copied over show `?` instead of number in the `Files`, `Dirs` and `Size` column. Newly created snapshots show actual numbers there.
  * Why is that?

**Reply by aawsome** [reply]
@mprasil Your source repo is a restic repository, right?
- rustic uses a much higher pack size for data packs, depending on the actual repository size
- if you copy from a snapshot, which doesn't have these metadata saved, it won't be saved in the target repository. What does `rustic snapshots` (using the profile you are using with `copy` show? The snapshots in the target repository should be identical. If your source snapshots have been generated by restic, they lack this information, this is WIP in restic.

**Reply by mprasil** [reply]
Yes exactly, the source is restic. Thanks for confirming.
